### ISSUE TITLE
Increasing thread count / RSS

### ISSUE THREAD

**Issue by irvingwashington** [GoogleCloudPlatform/google-cloud-ruby#1994]
Hey there, we’re observing a weird issue with the `google-cloud-pubsub` gem. 
On processes that start a subscriber, sleeping threads count increases without a limit.

The script below is enough for me to reproduce the issue, after 20 minutes the thread count is about 40, if I leave it for few hours it goes into hundreds. The highest count I've seen so far was over 3'000.

(subscriber.rb)
```ruby
require 'google/cloud/pubsub'

keyfile = 'path_to_the_keyfile'
project_id = 'correct_project_id'

topic_name = 'test_topic_name'
subscription_name = 'test-subscription'

@pubsub = pubsub = Google::Cloud::Pubsub.new(project_id: project_id, credentials: keyfile)

@topic = pubsub.topic(topic_name) || pubsub.create_topic(topic_name)
puts "Topic: #{@topic.inspect}"

@subscription = @topic.subscription(subscription_name) || t.create_subscription(subscription_name)
puts "Subscription: #{@subscription.inspect}"

@subscriber = @subscription.listen { |recv_msg| recv_msg.ack! } # Not really called

formatter = ->(sev, dt, _, msg) { "#{Thread.object_id} #{dt.utc.iso8601(3)} #{sev}: #{msg} (Threads num: #{Thread.list.size})\n" }

GRPC::DefaultLogger::LOGGER = Logger.new(STDOUT)
GRPC::DefaultLogger::LOGGER.formatter = formatter

@subscriber.start
```
I run it in irb to play with threads and ObjectSpace (`irb -r ./subscriber.rb`).

The following patch to GRPC seems to put stop to the issue, however I have too little context to know if it's the correct solution (the usage of GRPC might be wrong, also, I don't see anything like this in reported issues).

```
diff --git a/src/ruby/lib/grpc/generic/bidi_call.rb b/src/ruby/lib/grpc/generic/bidi_call.rb
index 3bdcc0062e..75d57744e7 100644
--- a/src/ruby/lib/grpc/generic/bidi_call.rb
+++ b/src/ruby/lib/grpc/generic/bidi_call.rb
@@ -208,6 +208,7 @@ module GRPC
       rescue StandardError => e
         GRPC.logger.warn('bidi: read-loop failed')
         GRPC.logger.warn(e)
+        @enq_th.kill
         raise e
       ensure
         set_input_stream_done.call
```

Some stack traces:
```
(Sleeping thread stack trace)
  /usr/local/bundle/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb:43:in `pop'
  /usr/local/bundle/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb:43:in `block in each'
  /usr/local/bundle/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb:42:in `loop'
  /usr/local/bundle/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb:42:in `each'
  /usr/local/bundle/gems/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/bidi_call.rb:139:in `each'
  /usr/local/bundle/gems/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/bidi_call.rb:139:in `each'
  /usr/local/bundle/gems/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/bidi_call.rb:139:in `each'
  /usr/local/bundle/gems/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/bidi_call.rb:139:in `each'
  /usr/local/bundle/gems/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/bidi_call.rb:139:in `write_loop'
  /usr/local/bundle/gems/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/bidi_call.rb:75:in `block in run_on_client’

(One of the paths that lead to that thread creation)
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/grpc-1.10.0-universal-darwin/src/ruby/lib/grpc/generic/bidi_call.rb:74:in `run_on_client'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/grpc-1.10.0-universal-darwin/src/ruby/lib/grpc/generic/active_call.rb:511:in `bidi_streamer'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/grpc-1.10.0-universal-darwin/src/ruby/lib/grpc/generic/client_stub.rb:473:in `block in bidi_streamer'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/grpc-1.10.0-universal-darwin/src/ruby/lib/grpc/generic/interceptors.rb:170:in `intercept!'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/grpc-1.10.0-universal-darwin/src/ruby/lib/grpc/generic/client_stub.rb:472:in `bidi_streamer'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/grpc-1.10.0-universal-darwin/src/ruby/lib/grpc/generic/service.rb:185:in `block (3 levels) in rpc_stub_class'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-gax-1.0.1/lib/google/gax/api_callable.rb:344:in `call'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-gax-1.0.1/lib/google/gax/api_callable.rb:344:in `block in retryable'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-gax-1.0.1/lib/google/gax/api_callable.rb:226:in `block in create_api_call'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-gax-1.0.1/lib/google/gax/api_callable.rb:256:in `block in create_api_call'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/v1/subscriber_client.rb:757:in `streaming_pull'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/service.rb:262:in `block in streaming_pull'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/service.rb:456:in `execute'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/service.rb:261:in `streaming_pull'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber/stream.rb:263:in `start_streaming!'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber/stream.rb:63:in `block in start'
  /Users/maciejdubinski/.rubies/ruby-2.4.3/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber/stream.rb:60:in `start'
  /Users/maciejdubinski/.gem/ruby/2.4.3/gems/google-cloud-pubsub-0.30.0/lib/google/cloud/pubsub/subscriber.rb:109:in `block (3 levels) in start’
```

The described behavior was observed on 0.29.0 and 0.30.0, on MRI 2.4.2 and 2.4.3, on macOS and on Debian.

So my questions are: is this a bug? If so, where does the bug belong to (here or GRPC)?

**Comment by blowmage** [comment]
@apolcyn Can you weigh in on this?

**Comment by apolcyn** [comment]
@irvingwashington thanks for the details and the stack trace of the blocked thread. Is it easy for you to also print out the exception that's occurring in the read loop?

e.g.,:

```
+++ b/src/ruby/lib/grpc/generic/bidi_call.rb
@@ -208,6 +208,7 @@ module GRPC
       rescue StandardError => e
         GRPC.logger.warn('bidi: read-loop failed')
         GRPC.logger.warn(e)
+         p e.inspect      
         raise e
```

?


From the looks of it, I'm guessing that the RPC is failing somehow and at the same time the writer thread (the thread running `write_loop`) of the bidi call is probably pulling on an enumerable that is still "alive" but also not getting requests or "sentinels" pushed into it. My off the cuff guess on a fix here is that the application should probably handle that exception and "signla an end to the request stream" by enqueuing a sentinel onto that request stream.

**Comment by blowmage** [comment]
I'm actively working on a couple of tests, but need to let them run for a while.

**Comment by blowmage** [comment]
@irvingwashington Thank you for this issue and the clear repro. It's clear and wonderfully written. I've been able to repro the slow increase in open threads using your script. I didn't let it run for very long, but I did see it slowly increase from what looks to be a baseline of 11 to 20 in about 10 minutes.

The first thing I did was modify your script to one that didn't use a Subscriber with its accompanying streams and thread pool.

```ruby
require "google/cloud/pubsub"

topic_name = "test_topic_name"
subscription_name = "test-subscription"

pubsub = Google::Cloud::Pubsub.new # Use env vars for initializer

topic = pubsub.topic(topic_name) || pubsub.create_topic(topic_name)
puts "Topic: #{topic.inspect}"

subscription = topic.subscription(subscription_name) || topic.create_subscription(subscription_name)
puts "Subscription: #{subscription.inspect}"

# Monkey-patch non-streaming listen
# No Subscriber, no concurrent-ruby pools, etc...
def subscription.non_streaming_listen max: 100, delay: 1
  loop do
    msgs = wait_for_messages max: max
    if msgs.any?
      msgs.each { |msg| yield msg }
    else
      sleep delay
    end
  end
end

# Set up logging messages
formatter = ->(sev, dt, _, msg) { "#{Thread.object_id} #{dt.utc.iso8601(3)} #{sev}: #{msg} (Threads num: #{Thread.list.size})\n" }

GRPC::DefaultLogger::LOGGER = Logger.new(STDOUT)
GRPC::DefaultLogger::LOGGER.formatter = formatter

# Call blocking method
subscription.non_streaming_listen { |recv_msg| recv_msg.ack! }
```

I let this run for ~30 minutes and it remained on 3 threads. This seems to be confirming @apolcyn's comment:

> From the looks of it, I'm guessing that the RPC is failing somehow and at the same time the writer thread (the thread running write_loop) of the bidi call is probably pulling on an enumerable that is still "alive" but also not getting requests or "sentinels" pushed into it.

We maintain a thread in `google-cloud-pubsub` for each stream's responses, and `grpc` maintains a thread for each stream's requests. One issue with using streaming calls is that they can fail and we need to recover. I assumed that the request thread also closes because an error is raised, but your patch shows that this is not the case. Each time a stream fails we make a new GRPC call and create a new response thread, so this all seems to confirm that the previous bidi request thread remains open.

To test this I made the following changes:

```diff
diff --git a/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb b/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb
index d837bc080..b6e343b97 100644
--- a/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb
+++ b/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/enumerator_queue.rb
@@ -30,9 +30,11 @@ module Google
             @queue.push obj
           end
 
-          def dump_queue
+          def quit_and_dump_queue
             objs = []
             objs << @queue.pop until @queue.empty?
+            # Signal that the enumerator is ready to end
+            @queue.push @sentinel
             objs
           end
 
diff --git a/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb b/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb
index 22782a356..c447d16d5 100644
--- a/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb
+++ b/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb
@@ -256,7 +256,7 @@ module Google
 
             # signal to the previous queue to shut down
             old_queue = []
-            old_queue = @request_queue.dump_queue if @request_queue
+            old_queue = @request_queue.quit_and_dump_queue if @request_queue
 
             @request_queue = EnumeratorQueue.new self
             @request_queue.push initial_input_request
```

With these changes I let your original script run for ~30 minutes. I observed that it remained fairly steady at 11 threads. (A process with a subscriber with a session pool of 4 has 3 baseline threads, 4 stream threads, and 4 bidi threads.) It would occasionally bump up to 12 threads, but then quickly drop back down to 11 threads. I am fairly sure was due to the failed streaming response recovery mentioned before.

Is it possible for you to also try these changes and let us know if you see the same behavior?

**Comment by irvingwashington** [comment]
Wow, thanks for timely answers, I wasn't ready for that 😄 

For the record, the read loop exception was 
`14:The service was unable to fulfill your request. Please try again. [code=8a75]`

Regarding the changes to `enumerator_queue.rb` and `stream.rb`, I ran them and it looks well. The thread count stays low (12) when idle. When publishing, it stays at 64 which I guess is the thread count of full pools for 4 streams, so seems fine.

**Comment by blowmage** [comment]
Agreed that 64 threads seems right for both publishing and subscribing. I'll try to get a release out with this fix today.

**Comment by blowmage** [comment]
This is released in google-cloud-pubsub 0.30.1
